cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java

```diff
@@ -269,65 +279,102 @@ protected Credential constructCredentialsFromRequest(final RequestContext contex
+            }
+            // Notify the OSF of the remote principal authentication.
+            final PrincipalAuthenticationResult remoteUserInfo
+                = notifyRemotePrincipalAuthenticated(credential, principal, PROTOCOL_CAS);
+            credential.setUsername(remoteUserInfo.getUsername());
+            credential.setInstitutionId(ticketGrantingTicket.getAuthentication().getPrincipal().getId());
             return credential;
         } else if (request.getParameter("username") != null && request.getParameter("verification_key") != null) {
-            // Construct credential if presented with (username, verification_key)
-            // This is used when:
-            //      1. User creates an account
-            //      2. User resets the password through forgot_password
-            //      3. User sets password when added as an unregistered contribution
-            // Note: Two-factor sign in works and remain unchanged
+            // Authentication through username and verification key.
             credential.setUsername(request.getParameter("username"));
             credential.setVerificationKey(request.getParameter("verification_key"));
             return credential;
+        } else {
+            return null;
         }
-
-        return null;
     }
 
     /**
      * Securely notify the OSF of a Remote Principal Authentication credential. Allows the OSF the opportunity
      * to create a verified user account and/or assign institutional affiliation to the user's account.
      *
      * @param credential the credential object bearing the authentication headers from the idp
+     * @param principal the principal constructed from external CAS authentication, or null if protocol is SAML
+     * @param protocol the protocol implementation used by remote authentication: Shibboleth (SAML), CAS (CAS)
      * @return the username from the idp and setup on the OSF
      * @throws AccountException a account exception
      */
-    private PrincipalAuthenticationResult notifyRemotePrincipalAuthenticated(final OpenScienceFrameworkCredential credential)
-            throws AccountException {
+    private PrincipalAuthenticationResult notifyRemotePrincipalAuthenticated(
+        final OpenScienceFrameworkCredential credential,
+        final Principal principal,
+        final String protocol
+    ) throws AccountException {
+
         try {
-            final JSONObject normalized = this.normalizeRemotePrincipal(credential);
-            final JSONObject provider = normalized.getJSONObject("provider");
+            final JSONObject normalizedPayload;
+            if (PROTOCOL_SAML.equals(protocol)) {
+                normalizedPayload = this.normalizeRemotePrincipal(credential);
+            } else if (PROTOCOL_CAS.equals(protocol)) {
+                normalizedPayload = this.normalizeRemotePrincipal((this.constructCredentialFromRemotePrincipal(credential, principal)));
```

Comment:
For CAS IDP, `credential` does not contain authenticated user's information yet. Call `constructCredentialFromRemotePrincipal` to update it with `principal` before normalization.